// react-api-data/lib/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useRequestHandler = exports.performApiRequest = exports.invalidateApiDataRequest = exports.purgeAllApiData = exports.getApiDataRequest = exports.configureApiData = exports.useActions = exports.useApiData = exports.setRequestHandler = exports.reducer = exports.purgeAll = exports.afterRehydrate = exports.invalidateRequest = exports.getDataFromTree = exports.getEntity = exports.getResultData = exports.getRequest = exports.performRequest = exports.configure = exports.withApiData = void 0;
var withApiData_1 = __importDefault(require("./withApiData"));
exports.withApiData = withApiData_1.default;
var configure_1 = require("./actions/configure");
Object.defineProperty(exports, "configure", { enumerable: true, get: function () { return configure_1.configure; } });
var afterRehydrate_1 = require("./actions/afterRehydrate");
Object.defineProperty(exports, "afterRehydrate", { enumerable: true, get: function () { return afterRehydrate_1.afterRehydrate; } });
var purgeAll_1 = require("./actions/purgeAll");
Object.defineProperty(exports, "purgeAll", { enumerable: true, get: function () { return purgeAll_1.purgeAll; } });
var getRequest_1 = require("./selectors/getRequest");
Object.defineProperty(exports, "getRequest", { enumerable: true, get: function () { return getRequest_1.getRequest; } });
var getResultData_1 = require("./selectors/getResultData");
Object.defineProperty(exports, "getResultData", { enumerable: true, get: function () { return getResultData_1.getResultData; } });
var invalidateRequest_1 = require("./actions/invalidateRequest");
Object.defineProperty(exports, "invalidateRequest", { enumerable: true, get: function () { return invalidateRequest_1.invalidateRequest; } });
var performRequest_1 = require("./actions/performRequest");
Object.defineProperty(exports, "performRequest", { enumerable: true, get: function () { return performRequest_1.performRequest; } });
Object.defineProperty(exports, "setRequestHandler", { enumerable: true, get: function () { return performRequest_1.setRequestHandler; } });
var getDataFromTree_1 = __importDefault(require("./getDataFromTree"));
exports.getDataFromTree = getDataFromTree_1.default;
var getEntity_1 = require("./selectors/getEntity");
Object.defineProperty(exports, "getEntity", { enumerable: true, get: function () { return getEntity_1.getEntity; } });
var reducer_1 = __importDefault(require("./reducer"));
exports.reducer = reducer_1.default;
var useApiData_1 = __importDefault(require("./useApiData"));
exports.useApiData = useApiData_1.default;
var useActions_1 = __importDefault(require("./useActions"));
exports.useActions = useActions_1.default;
/**
 * @deprecated
 */
exports.configureApiData = configure_1.configure;
/**
 * @deprecated
 */
exports.getApiDataRequest = getRequest_1.getRequest;
/**
 * @deprecated
 */
exports.purgeAllApiData = purgeAll_1.purgeAll;
/**
 * @deprecated
 */
exports.invalidateApiDataRequest = invalidateRequest_1.invalidateRequest;
/**
 * @deprecated
 */
exports.performApiRequest = performRequest_1.performRequest;
/**
 * @deprecated
 */
exports.useRequestHandler = performRequest_1.setRequestHandler;
